t/plugin/redirect.t

```diff
@@ -0,0 +1,288 @@
+            if not ok then
+                ngx.say(err)
+            end
+
+            ngx.say("done")
+        }
+    }
+--- request
+GET /t
+--- response_body
+done
+--- no_error_log
+[error]
+
+
+
+=== TEST 2: default ret_code
+--- config
+    location /t {
+        content_by_lua_block {
+            local plugin = require("apisix.plugins.redirect")
+            local ok, err = plugin.check_schema({
+                -- ret_code = 302,
+                uri = '/foo',
+            })
+            if not ok then
+                ngx.say(err)
+            end
+
+            ngx.say("done")
+        }
+    }
+--- request
+GET /t
+--- response_body
+done
+--- no_error_log
+[error]
+
+
+
+=== TEST 3: add plugin with new uri: /test/add
+--- config
+    location /t {
+        content_by_lua_block {
+            local t = require("lib.test_admin").test
+            local code, body = t('/apisix/admin/routes/1',
+                ngx.HTTP_PUT,
+                [[{
+                    "plugins": {
+                        "redirect": {
+                            "uri": "/test/add",
+                            "ret_code": 301
+                        }
+                    },
+                    "uri": "/hello"
+                }]]
+                )
+
+            if code >= 300 then
+                ngx.status = code
+            end
+            ngx.say(body)
+        }
+    }
+--- request
+GET /t
+--- response_body
+passed
+--- no_error_log
+[error]
+
+
+
+=== TEST 4: redirect
+--- request
+GET /hello
+--- response_headers
+Location: /test/add
+--- error_code: 301
+--- no_error_log
+[error]
+
+
+
+=== TEST 5: add plugin with new uri: $uri/test/add
+--- config
+    location /t {
+        content_by_lua_block {
+            local t = require("lib.test_admin").test
+            local code, body = t('/apisix/admin/routes/1',
+                ngx.HTTP_PUT,
+                [[{
+                    "plugins": {
+                        "redirect": {
+                            "uri": "$uri/test/add",
+                            "ret_code": 301
+                        }
+                    },
+                    "uri": "/hello"
+                }]]
+                )
+
+            if code >= 300 then
+                ngx.status = code
+            end
+            ngx.say(body)
+        }
+    }
+--- request
+GET /t
+--- response_body
+passed
+--- no_error_log
+[error]
+
+
+
+=== TEST 6: redirect
+--- request
+GET /hello
+--- response_headers
+Location: /hello/test/add
+--- error_code: 301
+--- no_error_log
+[error]
+
+
+
+=== TEST 7: add plugin with new uri: $uri/test/a${arg_name}c
+--- config
+    location /t {
+        content_by_lua_block {
+            local t = require("lib.test_admin").test
+            local code, body = t('/apisix/admin/routes/1',
+                ngx.HTTP_PUT,
+                [[{
+                    "plugins": {
+                        "redirect": {
+                            "uri": "$uri/test/a${arg_name}c",
```

Comment:
need test case to cover`"uri": "$bad_var/test",`

Reply:
did